src/cares_wrap.cc

```diff
@@ -1797,22 +1797,24 @@ static void Query(const FunctionCallbackInfo<Value>& args) {
 
   Local<Object> req_wrap_obj = args[0].As<Object>();
   Local<String> string = args[1].As<String>();
-  Wrap* wrap = new Wrap(channel, req_wrap_obj);
+  std::unique_ptr<Wrap> wrap {new Wrap(channel, req_wrap_obj)};
 
   node::Utf8Value name(env->isolate(), string);
   channel->ModifyActivityQueryCount(1);
   int err = wrap->Send(*name);
   if (err) {
     channel->ModifyActivityQueryCount(-1);
-    delete wrap;
+  } else {
+    wrap.release();
   }
 
   args.GetReturnValue().Set(err);
 }
 
 
 void AfterGetAddrInfo(uv_getaddrinfo_t* req, int status, struct addrinfo* res) {
-  GetAddrInfoReqWrap* req_wrap = static_cast<GetAddrInfoReqWrap*>(req->data);
+  std::unique_ptr<GetAddrInfoReqWrap> req_wrap {
```

Comment:
This should be managed with an RAII wrapper not `unique_ptr`.
```suggested
  Wrap* wrap = new Wrap(channel, req_wrap_obj);
  OnScopeLeave([wrap](){ delete wrap; })
```
... and later
```cpp
if (err == 0) {
  OnScopeLeave.cancel()
}
```

Reply:
Or change the signature to:
```cpp
void AfterGetAddrInfo(uv_getaddrinfo_t* req, int status,
                      std::unique_ptr<GetAddrInfoReqWrap> req_wrap)
```
to express that this function take ownership of `req->data`

Reply:
> This should be managed with an RAII wrapper not `unique_ptr`.

So IMHO the right pattern here is [`ScopeGuard`](http://www.drdobbs.com/cpp/generic-change-the-way-you-write-excepti/184403758?pgno=3) but since we don't have an implementation for it yet, I agree this is a good aprouch.